tools/poetry.sh

```diff
@@ -45,13 +45,22 @@ install-poetry() {
     # Download the automated installer:
     installer=$poetry_home/install-poetry.py
     download-file --uri=https://install.python-poetry.org --out="$installer"
+    python_binary="$(find-python)"
     # Run the install:
     with-lock "$POETRY_HOME/.install.lock" \
         env POETRY_HOME="$poetry_home" \
-        "$(find-python)" -u "$installer" --yes --version "$poetry_version" \
+        "$python_binary" -u "$installer" --yes --version "$poetry_version" \
     || (
         cat -- poetry-installer*.log && fail "Poetry installation failed"
     )
+    # Extra step must be taken to ensure Poetry's virtual environment continues to use the correct Python binary.
+    # See: https://github.com/python-poetry/poetry/issues/522#issuecomment-492369712
+    with-lock "$POETRY_HOME/.install.lock" \
+        env POETRY_HOME="$poetry_home" \
+        "$POETRY_EXE" env use "$python_binary" \
+    || (
+        fail "Poetry installation failed (env use <path/to/python>)"
+    )
```

Comment:
Having a separate `with-lock` is safe, as the `env use` subcommand applies to only the current project (from `pwd` or from `-C`), so concurrent invocations in separate project directories will not contend. It also runs very quickly as all it does is splat some files down.

This also means, however, that this step needs to run any time that Poetry will create a new virtualenv, and not just once when it is installed. Because it's fast, idempotent, and cheap, it would suffice to run `poetry env use --quiet <python>` before every Poetry command.